Sprint 5/step 1

maybe777 opened the pull request:
Ошибка типизации dispatch

maybe777 committed:
Попытка сдачи 4 спинта.

maybe777 committed:
Исправление замечаний ревьюера (1-я итерация)

maybe777 committed:
Почти тимизирован Redux. Проблемы с UserProfile.

maybe777 committed:
Почти типизирован Redux. Проблемы с UserProfile. Нужно избавиться от типа any в user-profile-reducer.tsx

maybe777 committed:
Временнный фикс - тип AppThunk убран из всех Actions.

Починено!

maybe777 committed:
Получилось настроить middleware для работы с feed.

maybe777 committed:
Готова лента заказов без роутинга.

maybe777 committed:
Готова лента заказов без роутинга.

maybe777 committed:
Реализована ветка сценария feed.

maybe777 committed:
Реализован профиль пользователя, и все роутинги. Осалось добавить еще один action для получения заказа по user.

maybe777 committed:
Функционал готов. Попытка сдачи (1-я итерация)

gennady-bars commented on src/service/hooks.ts:
```diff
@@ -0,0 +1,12 @@
+export const useSelector: TypedUseSelectorHook<TRootState> = selectorHook;
+
+// Хук не даст отправить экшен, который ему не знаком
+export const useDispatch = () => dispatchHook<TAppDispatch>();
```
Отлично типизировали хуки `useSelector` и `useDispatch`

gennady-bars commented on src/redux/types/index.tsx:
```diff
@@ -0,0 +1,23 @@
+import rootReducer from "../reducers/root-reducer";
+import {TWSOrderActions} from "./ws-order-action-types";
+
+export type TRootState = ReturnType<typeof rootReducer>;
```
Отлично типизировали стейт, диспатч и `thunk`

gennady-bars commented on src/redux/middleware/socket-middleware.ts:
```diff
@@ -0,0 +1,65 @@
+import {TAppDispatch, TRootState} from "../types";
+
+
+export const socketMiddleware = (wsActions: any): Middleware => {
```
Отличный универсальный `socketMiddleware` сделали

gennady-bars replied:
объект с типами экшенов нужно хотя бы типизировать так, чтобы это был объект: `{[key:string]: any}`


```js

export const socketMiddleware = (wsActions: {[key:string]: any}): Middleware => {


```

gennady-bars commented on src/redux/middleware/socket-middleware.ts:
```diff
@@ -0,0 +1,65 @@
+
+            if (type === wsInit) {
+                url = action.payload
+                socket = new WebSocket(url)
```
Отлично вставляете `url` в сокет. Очень универсально

gennady-bars commented on src/components/app/app.tsx:
```diff
@@ -12,28 +12,39 @@ import styles from "./app.module.css";
     useEffect(() => {
-        // @ts-ignore
         dispatch(getData())
+        dispatch(wsConnection(WS_FEED_URL))
```
Открывать соединение сокетов нельзя в `App`, так как это лишняя нагрузка на сервер. При входе на страницу ленты заказов нужно открывать соединение сокета, а при покидании страницы – закрывать. Это делается все в одном хуке `useEffect`. То же самое и с историей заказов

maybe777 replied:
Если так сделать - отваливается функциональность открытия деталей заказа вне контекста страницы, согласно задания.

Если пользователь попадает на маршруты /feed/:id и /profile/orders/:id мануально, то есть обновляет текущую страницу или копирует ссылку на заказ, то у него откроется не модальное окно, а детальная страница «Информация о заказе»:

gennady-bars replied:
> Если так сделать - отваливается функциональность открытия деталей заказа вне контекста страницы, согласно задания.
> 
> Если пользователь попадает на маршруты /feed/:id и /profile/orders/:id мануально, то есть обновляет текущую страницу или копирует ссылку на заказ, то у него откроется не модальное окно, а детальная страница «Информация о заказе»:


Нужно открывать соединение сокетов и там, где страница заказа отображается тоже. При вставке ссылки в браузер, будет тоже открываться, а при уходе со страницы с деталями заказа, закрываться


А держать сразу 2 открытых соединения не стоит. Когда нужно, тогда открывается, а при уходе со страниц - закрывать нужно, чтобы зря не нагружать сервер и приложение тоже (оно может притормаживать из-за этого)

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Работа проделана огромная

* Отлично типизировали хуки `useSelector` и `useDispatch`
* Отлично типизировали стейт, диспатч и `thunk`
* Отличный универсальный `socketMiddleware` сделали
* Хорошо типизировали редакс и проект в целом


но есть некоторые недочеты:


* при входе в ленту и историю заказов в консоли ошибка `key`. Скрин http://joxi.ru/DrlW6XVSK8Le5A 
* Открывать соединение сокетов нельзя в `App`, так как это лишняя нагрузка на сервер. При входе на страницу ленты заказов нужно открывать соединение сокета, а при покидании страницы – закрывать. Это делается все в одном хуке `useEffect`. То же самое и с историей заказов
* объект с типами экшенов нужно хотя бы типизировать так, чтобы это был объект: `{[key:string]: any}`
* Вам больше не нужно типизировать аргументы в `map`, `filter`, `reduce` и тд, если массив уже 1 раз типизирован. Тайпскрипт возьмет тип автоматически. Код станет чище



**Можно лучше**

*  желательно очищать конструктор после успешного получения номера заказа с сервера в блоке `then`, чтобы пользователь мог следующий заказ сделать, не удаляя старые ингредиенты
* лучше отображать эффект **загрузки** при ожидании ответа от сервера с номером заказа.


Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте работоспособность проекта и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

Удачного рефакторинга кода.

maybe777 committed:
-- при входе в ленту и историю заказов в консоли ошибка key. Скрин http://joxi.ru/DrlW6XVSK8Le5A
Поправил

--Открывать соединение сокетов нельзя в App, так как это лишняя нагрузка на сервер. При входе на страницу ленты заказов нужно открывать соединение сокета, а при покидании страницы – закрывать. Это делается все в одном хуке useEffect. То же самое и с историей заказов
Убрав из App получение данных - теряю функциональность, описанную в задании = (Если пользователь попадает на маршруты /feed/:id и /profile/orders/:id мануально, то есть обновляет текущую страницу или копирует ссылку на заказ, то у него откроется не модальное окно, а детальная страница «Информация о заказе»:)

--объект с типами экшенов нужно хотя бы типизировать так, чтобы это был объект: {[key:string]: any}
Вам больше не нужно типизировать аргументы в map, filter, reduce и тд, если массив уже 1 раз типизирован. Тайпскрипт возьмет тип автоматически. Код станет чище
Готово

----Можно лучше
--желательно очищать конструктор после успешного получения номера заказа с сервера в блоке then, чтобы пользователь мог следующий заказ сделать, не удаляя старые ингредиенты
Сделал - при закрытии модального окна с заказом

--лучше отображать эффект загрузки при ожидании ответа от сервера с номером заказа.

gennady-bars reviewed (commented):
Еще немного и все будет отлично:

* Нужно открывать соединение сокетов и там, где страница заказа отображается тоже. При вставке ссылки в браузер, будет тоже открываться, а при уходе со страницы с деталями заказа, закрываться
* при входе в историю ошибка `key`. Скрин http://joxi.ru/vAW8b9ZT3opEZr 
* иногда при входе в историю ломается приложение (видимо, нужно проверять, пришел ли ответ с заказами)
* при открытии попапа с заказами в консоли ошибка `key`  тоже. Скрин http://joxi.ru/4Ak65neTj7We62

maybe777 committed:
3-я попытка сдачи

maybe777 committed:
3-я попытка сдачи

gennady-bars reviewed (commented):
Уже почти все отлично, но:

* все еще ошибки `key` в ленте и истории. Скрин http://joxi.ru/1A5YdBpS46n4aA . И при открытии попапов там же тоже ошибка `key`.  Это потому, что `key` там нужно указывать в компоненте `Link`, так как он является самым верхним уровнем верстки. Скрин http://joxi.ru/4Ak65neTj8yVN2 . Нужно везде посмотреть, где указали `key`, и переместить его на самый верх
* при вставке ссылки в браузер не отображаются заказы. Нужно на страницах заказов тоже открывать и закрывать соединениях

maybe777 committed:
4-я попытка сдачи

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
